### ISSUE TITLE
Fix layout not receiving template params when using zend/renderer with zend-mvc

### ISSUE THREAD

**Issue by miv23** [acelaya/ZF-AcMailer#200]
Settings are in mail.global.php and I have stored the template file in my view folder under email/contact.phtml. The following code results in a "can't resolve to file" error:

```
return [
  'acmailer_options' => [

    /*
    * This is where you configure all the emails that can be sent by the application.
    * You can create new ones at runtime, but this eases predefining emails that do not need to 
      change
    */
    'emails' => [
      'contact' => [
        'template' => 'application/email/contact',
```

I have also tried just using the layout param in the 'template_params' array when calling the send method, but without specifying a template it is ignored. Is it possible to specify the template when calling the send method?

**Comment by acelaya** [comment]
The template has to follow the same syntax as any other template in your application. If you can render the 'application/email/contact' template in a controller's action, it should also work in the case of emails. 

In this case, in order to get it working, you have to have the "views" folder of your module, containing an "application" folder, containing an "email" folder, containing the "contact.phtml" file.

However, if it is not working, it would be helpful to be able to have the error you are getting.

The template can also be set dynamically at runtime. Just remove the param in the configuration, and when calling `send`, pass a second argument like this:

```php
<?php
$mailService->send('contact', [
    'template' => 'application/email/contact',
]);
```

The layout is different from the template, and can be reused among many templates (as in regular MVC views).

If you don't want to use a layout, just ignore it. It is not required.

**Comment by miv23** [comment]
Ok I've got the template working. My mistake was thinking I didn't need to set up the route and action which I now have using the IndexController. 

It works if I set the template in mail.global.php or dynamically when calling send, but only if I also specify the layout param in template_params. Without layout I get an error saying the view_model is empty. So both template and layout have the same path - application/index/contact.

I just have one other problem. I'm trying to send a variable to the template, but I'm having trouble accessing it. I'm sending it via the template_params array:

```
'template_params' => [
                        'layout' => 'application/index/contact',
                        'email' => $email,
                    ],
``` 

I can verify that $email has a value as it is stored in the db.

I've tried many variations to try access this variable in my template. The first one I tried which I expected to work was $this->email, but no matter what I've tried the value is always blank. 

Am I using template_params incorrectly? 
Thanks for the help so far. :)

**Comment by acelaya** [comment]
It shouldn't be necessary to pass the layout. Even less, having to pass the same view as the template and the layout.

Regarding the param problem, my guess is that it is rendering the layout only, while the arguments are only passed to the template. That's why you can't get it working.

I will look into both problems. For example, it might make sense to pass the template params to the layout too.

I'll investigate and let you know.

Thanks for your patience :)

**Comment by juuce** [comment]
Hi, I too am confused with this new format/structure of view/template definition and passing parameters.
Firstly, thanks for the great module. Been using it for a couple of projects and up to v6 I haven't had any major issues. 

I have additional questions to the above:
1. How can I add a variable into the layout template using the ->send([ .. ]) format? My reason for this is linking to external logos and styling (eg. url). Currently I cannot seem to do this with the v7 format.
2. Same issue as above - I am only seeing the main layout template and cannot get the child template to render....
e.g. in template from config
`[ 'view_manager' => ['template_map' =>
 'mail/start' => _DIR__.'/../view/mail/start.phtml' ]
 'mail/template' => _DIR__.'/../view/mail/layout_template.phtml' ]
]`

```
BASE URL: <?php  echo $this->baseUrl; ?>
---- CONTENT SHOULD BE BELOW ---
<?php echo $this->template;?>
```
Parameters to pass to send()
```
$mainTemplate = 'mail/template';
$bodyTemplate = 'mail/start';
$params = [
            'to'       => [$toEmail],
            'template' => $bodyTemplate,
            'template_params' => [
                'title' => 'Welcome to CareerMagnifier',
                'baseUrl' => $this->baseUrl,
                'layout' => $mainTemplate
            ]
];
```

This results in the email (Testing to file as transport for now) with the main layout template but no variables echoed and no 'inner/child' template being rendered.
How does the parameter definition know which child should be attached to which variable? (e.g. template)

It seems that flexibility has somewhat been reduced and I am likely going to revert to v6 for now... I'm sure this can be worked out in the future and looking forward to using your module as v7 in the future.

**Comment by acelaya** [comment]
There seems to be two different problems here.

The first one seems to be the fact that params are being passed to the inner template only, and not the layout, but that should be easy to fix.

The second one is that, as far as I can see, you are trying to render the inner template using the "template" name, while you should use "content" instead. That is not configurable right now and it might be better documented.

The problem here is that I'm not going to be able to fix this in the near future (at least in the next two weeks), so if any of you want to give it a try, a PR would be very appreciated.

**Comment by juuce** [comment]
Thanks, 
Confirmed that Child template is statically linked to 'content'
i.e. in Main Layout file this will render the child + also parameters render correctly (baseUrl)
`<?= $this->content;?>`

I'll try and provide an update for adding params to the main layout + possibly allow multiple child templates or at least a variable from the default 'content'. Probably overkill for email but see how it goes. No guarantee i'll get it going :)

**Comment by juuce** [comment]
Hi, Never performed a PR. Tried to clone AcMailer and upload a new branch to compare... No permissions to upload.

New parameters to pass into template_params
- child_template_name - string - name of variable defaults to 'content' as per current
- layout_params - bool|array - if bool passed all child params passed to main layout, if array passed only those added to template

Code I've used works well and is below. (in file AcMailer/src/View/MvcMailViewRenderer.php)

```
public function render(string $name, array $params = []): string
    {
        $layout = $params['layout'] ?? null;
        unset($params['layout']);
        
        $viewModel = new ViewModel($params);
        $viewModel->setTemplate($name);
        
        // If a layout was provided, add the original view model as a child
        if ($layout !== null) {
            
            $childTemplateName = $params['child_template_name'] ?? 'content';
            $layoutModelParams = [
                $childTemplateName => $this->renderer->render($viewModel)
            ];
            
            if ( isset($params['layout_params']) ) {
                if ( is_bool($params['layout_params'] && $params['layout_params'] === true) ) {
                    $layoutModelParams = array_merge($layoutModelParams, $params);
                }
                if ( is_array($params['layout_params']) ) {
                    $layoutModelParams = array_merge($layoutModelParams, $params['layout_params']);
                }
            }
            
            $layoutModel = new ViewModel($layoutModelParams);
            $layoutModel->setTemplate($layout);
            $viewModel = $layoutModel;
        }
        
        return $this->renderer->render($viewModel);
    }
```

**Comment by acelaya** [comment]
You have to fork the repository, which will "copy" it under your username. Then, create a feature branch there, perform changes, and then GitHub will allow you to create a PR from your repo to this one.

**Comment by juuce** [comment]
Are you OK with those additional parameters? AcMailer\Model\Email::setTemplateParameters();

1. child_template_name: string
2. layout_params: bool|string[]

If not please provide any changes to the above code

**Comment by acelaya** [comment]
I wouldn't define two sets of parameters, but rather propagate all the template params to the layout too. That would make the behavior consistent with how all the other renderers work.

On the other hand, the child template name might make sense as long as you ensure the "content" name is used of not defined, to make it backwards compatible.

**Comment by juuce** [comment]
Sorry, already created pull request. You can adjust how you like. I think your ZF knowledge is better than mine so I'll leave it to you.

**Comment by acelaya** [comment]
If you keep pushing changes to your branch, the PR gets updated, so you can apply as many changes as you want before getting it merged.

The usual workflow is that you create the PR, the repo owner reviews it and proposes changes if any, and then you implement those changes and update the PR.

**Comment by acelaya** [comment]
This has already been fixed and released in v7.2.1

Now params are properly propagated both to the layout and the template when working with Zend MVC. Also the child template "capture to" can be configured.

Thanks you @miv23 and @juuce for your feedback :)
